macro/src/argument.rs

```diff
@@ -89,17 +89,93 @@ impl Parse for JavaClass {
     }
 }
 
-pub enum MemberListing {
-    All,
+#[derive(Clone, Debug)]
+pub struct MemberListing {
+    elements: Vec<MemberListingElement>,
+}
+
+#[derive(Clone, Debug)]
+pub enum MemberListingElement {
+    /// `* - <ML>`
+    Wildcard(MemberListing),
+
+    /// `void toString()` -- full details
+    Named(SpannedMethodSig),
+}
+
+impl MemberListing {
+    pub fn contains_method(&self, m: &Method) -> bool {
+        self.elements.iter().any(|e| e.contains_method(m))
+    }
+
+    pub fn contains_constructor(&self, class: &ClassInfo, ctor: &Constructor) -> bool {
+        self.elements
+            .iter()
+            .any(|e| e.contains_constructor(class, ctor))
+    }
+
+    pub fn all() -> Self {
+        MemberListing {
+            elements: vec![MemberListingElement::Wildcard(MemberListing::none())],
+        }
+    }
+
+    pub fn none() -> Self {
+        MemberListing { elements: vec![] }
+    }
+}
+
+impl MemberListingElement {
+    pub fn contains_method(&self, m: &Method) -> bool {
+        match self {
+            MemberListingElement::Wildcard(ml) => !ml.contains_method(m),
+            MemberListingElement::Named(n) => n.method_sig.matches(m),
+        }
+    }
+
+    pub fn contains_constructor(&self, class: &ClassInfo, ctor: &Constructor) -> bool {
+        match self {
+            MemberListingElement::Wildcard(ml) => !ml.contains_constructor(class, ctor),
+            MemberListingElement::Named(n) => n.method_sig.matches_constructor(class, ctor),
+        }
+    }
 }
 
 impl Parse for MemberListing {
     fn parse(p: &mut Parser) -> Result<Option<Self>, SpanError> {
+        Ok(Some(MemberListing {
+            elements: MemberListingElement::parse_many(p)?,
+        }))
```

Comment:
Would this allow for mixing `*` and `method();` declarations together? Would we want to limit this to be either a wild card or an explicit method listing?

Reply:
Yes, technically it does, and there isn't much point to it. It probably makes sense to limit it, but it's also kind of...meh...it's harmless? I guess it might block us from doing other things in the future.

Reply:
I'm inclined to add a FIXME and maybe address it as part of a proper LALRPOP grammar in the future.

Reply:
Covered under #31